libsolidity/ast/AsmJsonConverter.cpp

```diff
@@ -0,0 +1,272 @@
+	ret["nodeType"] = _nodeType;
+	int length = -1;
+	if (_location.start >= 0 && _location.end >= 0)
+		length = _location.end - _location.start;
+	ret["src"] = to_string(_location.start) + ":" + to_string(length) + ":" + to_string(m_sourceIndex);
+	return ret;
+}
+
+template <class T>
+Json::Value AsmJsonConverter::vectorOfVariantsToJson(vector<T> const& _vec ) const {
+	Json::Value ret(Json::arrayValue);
+	// TODO check what is returned when _vec is empty? nullPtr or []
+	for (auto const& var: _vec) {
+		ret.append(boost::apply_visitor(*this, var));
+	}
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::Block const& _node) const
+{
+	Json::Value ret = createAstNode(_node.location, "Block");
+	ret["statements"] = vectorOfVariantsToJson(_node.statements);
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::TypedName const& _node) const
+{
+	solAssert(!_node.name.empty(), "Invalid variable name.");
+	Json::Value ret = createAstNode(_node.location, "TypedName");
+	ret["name"] = _node.name.str();
+	ret["type"] = _node.type.str();
+	//NOTE isn't it weird that both are YulStrings but they are referenced differently in AsmData.h?
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::Literal const& _node) const
+{
+	Json::Value ret = createAstNode(_node.location, "Literal");
+	switch (_node.kind)
+	{
+	case yul::LiteralKind::Number:
+		solAssert(isValidDecimal(_node.value.str()) || isValidHex(_node.value.str()), "Invalid number literal");
+		ret["type"] = _node.type.str(); //_node.value.str() + appendTypeName(_node.type);
+		break;
+	case yul::LiteralKind::Boolean:
+		// ?? what does this notation even mean >"true"_yulstring< ?
+//		solAssert(_node.value == "true"_yulstring || _node.value == "false"_yulstring, "Invalid bool literal.");
+		ret["type"] =  "Boolean"; //NOTE should it be like this? : ((_node.value == "true"_yulstring) ? "true" : "false") + appendTypeName(_node.type);
+		break;
+	case yul::LiteralKind::String:
+		ret["type"] = "String";
+		break;
+	default:
+		solAssert(false, "unknown type of literal");
+	}
+
+	// NOTE: in AsmPrinter, there happens a lot more processing before the string is printed
+	// Do I need to do the same?
+	ret["value"] = _node.value.str();
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::Identifier const& _node) const
+{
+	solAssert(!_node.name.empty(), "Invalid identifier");
+	Json::Value ret = createAstNode(_node.location, "Identifier");
+	ret["name"] = _node.name.str();
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::Assignment const& _node) const
+{
+	solAssert(_node.variableNames.size() >= 1, "Invalid assignment syntax");
+	Json::Value ret = createAstNode(_node.location, "Assignment");
+
+	// TODO write helper function that applies (*this) to every element of vector
+	Json::Value varNames(Json::arrayValue);
+	for (auto const& var: _node.variableNames)
+		varNames.append((*this)(var));
+	ret["variableNames"] = std::move(varNames);
+
```

Comment:
I think the above could be done in a smarter way... any hints what a helper could look like?